### PR TITLE
Add support for the Falcon new decoder architecture

### PR DESCRIPTION
**Pull request by danieldk**
## Description

The 40B Falcon model uses the so-called new decoder architecture. This change adds support for the new decoder architecture. This necessitates a bunch of changes across the board:

* So far we supported having a uniform number of query/key/value heads or full head-sharing of key/value heads. The new decoder architecture provides a configurable number of key/value heads, where the number of query heads is a multiple of the number of key/value heads. In order to support this, we replace the `QkvHeadSharing` enum by a `AttentionHeads` class that allows more flexible configurations. The attention layer is extended to support this new scenario.

* The new decoder architecture's transformer layer is much more canonical, allowing us to reuse the shared decoder layer. However, in contrast to the other decoders that use the shared layer, Falcon puts the dropout after parallel attention. To accomodate more flexible dropouts configurations, we introduce the `TransformerDropouts` class, which works similar to the `TransformerLayerNorms` class, but for dropouts.

* Split the HF configuration parsing for Falcon in functions for the `RefinedWebModel` and `falcon` model types.

This change also adds two new models to test the new-decoder architecture.

### Types of change

Feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/models/falcon/config.py [hunk 1]
```diff
@@ -23,8 +23,8 @@ def __init__(
         *,
         dropout_prob: float = 0.1,
         hidden_width: int = 2560,
-        multi_query: bool = True,
-        num_attention_heads: int = 71,
+        num_query_heads: int = 71,
+        num_kv_heads: int = 1,
```
Comment (shadeMe) [inline curated_transformers/models/falcon/config.py:27]: Why `kv` and not `key_value`? We're mapping them anyway.

File: docs/source/building-blocks.rst [hunk 2]
```diff
@@ -26,7 +26,7 @@ These modules and their helper classes implement the Transformer attention mecha
 .. autoclass:: curated_transformers.layers.attention.QkvMode
    :members:
 
-.. autoclass:: curated_transformers.layers.attention.QkvHeadSharing
+.. autoclass:: curated_transformers.layers.attention.AttentionHeads
```
Comment (shadeMe) [inline docs/source/building-blocks.rst:29]: Missing `TransformerDropouts` in the previous section.

File: curated_transformers/layers/attention.py [hunk 3]
```diff
@@ -138,19 +138,76 @@ def create_causal_mask(query: Tensor, key: Tensor) -> AttentionMask:
...
 
-    #: Multi-query attention: Key shares heads, value shares heads,
-    #: query has separate heads (`Shazeer et al., 2019`_).
-    #:
-    #: .. _Shazeer et al., 2019: https://arxiv.org/abs/1911.02150
-    KEY_VALUE = 1
+    @classmethod
+    def multi_query(cls, num_query_heads: int) -> "AttentionHeads":
+        """
+        Construct a multi-query attention configuration: key shares heads,
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/layers/attention.py]: I think the sharing aspect of it is confusing, I rewrote it to be more explicit.

File: curated_transformers/layers/attention.py [hunk 4]
```diff
@@ -505,6 +570,53 @@ def split_heads(input: Tensor, num_heads: int) -> Tensor:
...
+def split_heads_grouped_qkv(
+    projection: Tensor, attention_heads: AttentionHeads, dims_per_head: int
+) -> Tuple[Tensor, Tensor, Tensor]:
+    """
+    Split attention heads by grouping the heads in blocks of
+    the number of key/value-heads. Then distribute the heads
+    across query, key, and value.
+
+    :param projection:
+        The fused query, key, value projection.
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/layers/attention.py:586]: Added.

File: curated_transformers/layers/attention.py [hunk 5]
```diff
@@ -505,6 +570,53 @@ def split_heads(input: Tensor, num_heads: int) -> Tensor:
...
+    :param attention_heads:
+        Attention head configuration.
+    :param dims_per_head:
+        Head dimensionality.
+    :returns:
+        Query, key, value tensors.
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/layers/attention.py:595]: Added.

File: curated_transformers/layers/transformer.py [hunk 6]
```diff
@@ -34,6 +34,54 @@ class TransformerLayerNorms:
...
+class TransformerDropouts:
+    """
+    Dropouts used in a transformer layer.
+
+    By default, all the dropouts are disabled by setting the dropout
+    to the Torch ``Identity`` module. Therefore, only dropouts that are
+    needed have to be set.
+    """
+
+    #: Dropout of the output of the attention layer.
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/layers/transformer.py]: Fixed.

File: curated_transformers/models/falcon/config.py (hunk 1 above)
Comment (danieldk) [reply curated_transformers/models/falcon/config.py]: I renamed them later, forgot these. Fixed.

File: docs/source/building-blocks.rst (hunk 2 above)
Comment (danieldk) [reply docs/source/building-blocks.rst:32]: Added.

File: curated_transformers/layers/attention.py [hunk 7]
```diff
@@ -728,12 +734,20 @@ def split_heads_grouped_qkv(
...
+        *Shape:* ``(batch_size, seq_len,
+        (num_query_heads + 2*num_key_value_heads)*dims_per_head)``
     :param attention_heads:
         Attention head configuration.
     :param dims_per_head:
         Head dimensionality.
     :returns:
         Query, key, value tensors.
+
+        Shapes:
```
Comment (shadeMe) [inline curated_transformers/layers/attention.py:747]:
```suggestion
        *Shapes:*
```

**Review by shadeMe — APPROVED** [review]
LGTM! One minor fix.

**Pull request merged**